src/linear_algebra/annihilating_polynomial.lean

```diff
@@ -0,0 +1,197 @@
+      apply h, }, },
+end
+
+/-- The annihilating ideal generator is a member of the annihilating ideal,
+  following submodule.generator_mem -/
+lemma ann_ideal_generator_mem (a : A) : ann_ideal_generator 𝕜 a ∈ ann_ideal 𝕜 a :=
+let I := ann_ideal 𝕜 a,
+    g := submodule.is_principal.generator I in
+  I.mul_mem_left (C g.leading_coeff⁻¹) (submodule.is_principal.generator_mem I)
+
+/-- sourced from submodule.mem_iff_eq_smul_generator -/
+lemma mem_iff_eq_smul_ann_ideal_generator {p : 𝕜[X]} (a : A) :
+  p ∈ ann_ideal 𝕜 a ↔ ∃ s : 𝕜[X], p = s • ann_ideal_generator 𝕜 a :=
+by simp_rw [@eq_comm _ p, ← mem_span_singleton, span_singleton_ann_ideal_generator]
+
+/-- sourced from submodule.eq_bot_iff_generator_eq_zero -/
+lemma eq_bot_iff_ann_ideal_generator_eq_zero (a : A) :
+  ann_ideal 𝕜 a = (⊥ : ideal 𝕜[X]) ↔ ann_ideal_generator 𝕜 a = (0 : 𝕜[X]) :=
+begin
+  rw ← span_singleton_ann_ideal_generator,
+  apply @span_singleton_eq_bot 𝕜[X] 𝕜[X] _ _ _ (ann_ideal_generator 𝕜 a),
+end
+
+/-- The generator we chose for the annihilating ideal is monic when the ideal is non-zero. -/
+lemma monic_of_ann_ideal_generator (a : A) (hg : (ann_ideal_generator 𝕜 a : 𝕜[X]) ≠ 0) :
+  monic (ann_ideal_generator 𝕜 a : 𝕜[X]) :=
+begin
+  dunfold ann_ideal_generator,
+  dsimp *,
+  rw mul_comm,
+  have hg' : is_principal.generator (ann_ideal 𝕜 a) ≠ 0,
+  { unfold ann_ideal_generator at hg, simp at hg, exact hg, },
+  apply polynomial.monic_mul_leading_coeff_inv,
+  apply hg',
+end
+
+/-- We are working toward showing the generator of the annihilating ideal
+in the field case is the minimal polynomial. We are going to use a uniqueness
+theorem of the minimal polynomial. This is the first condition: it must annihilate
+the original element `a : A`. -/
+lemma ann_ideal_generator_aeval_eq_zero (a : A) :
+  aeval a (ann_ideal_generator 𝕜 a : 𝕜[X]) = 0 :=
+begin
+  have hg : aeval a (is_principal.generator (ann_ideal 𝕜 a)) = 0,
+  { have gen_member := submodule.is_principal.generator_mem (ann_ideal 𝕜 a),
+    exact (ring_hom.mem_ker (polynomial.aeval a).to_ring_hom).1 gen_member, },
+  rw ann_ideal_generator, simp *,
+end
+
+/-- sourced from submodule.is_principal.mem_iff_generator_dvd -/
+lemma mem_iff_ann_ideal_generator_dvd (a : A) {x : 𝕜[X]} :
+  x ∈ ann_ideal 𝕜 a ↔ ann_ideal_generator 𝕜 a ∣ x :=
+(mem_iff_eq_smul_ann_ideal_generator 𝕜 a).trans
+ (exists_congr (λ a, by simp only [mul_comm, smul_eq_mul]))
+
+/-- The generator of the annihilating ideal has minimal degree among
+ the non-zero members of the annihilating ideal -/
+lemma degree_ann_ideal_generator_le_of_mem (a : A) (p : 𝕜[X])
+  (hp : p ∈ ann_ideal 𝕜 a) (hpn0 : p ≠ 0) :
+  degree (ann_ideal_generator 𝕜 a : 𝕜[X]) ≤ degree p :=
+degree_le_of_dvd ((mem_iff_ann_ideal_generator_dvd 𝕜 a).1 hp) hpn0
+
+/-- This is what we have been building to:
+The monic generator of the annihilating ideal is the minimal polynomial. -/
+lemma minpoly_eq_monic_ann_ideal_generator (a : A) :
+  ann_ideal_generator 𝕜 a = minpoly 𝕜 a :=
```

Comment:
This actually suggests that `ann_ideal_generator` is kind of a useless definition: since the hypotheses are exactly the same as those for the definition of `ann_ideal_generator`, we might as well just replace `ann_ideal_generator` with `minpoly` everywhere.

Instead, the interesting part is that if `p` is any monic polynomial that generates `ann_ideal`, then it must be the minimal polynomial. So a better way to express this theorem would be:
```suggestion
lemma minpoly_eq_monic_ann_ideal_generator (a : A) (p : 𝕜[X])
  (p_monic : p.monic) (p_gen : ideal.span {p} = ann_ideal 𝕜 a) :
  minpoly 𝕜 a = p :
```
And then another useful theorem would state that for any nonzero ideal of polynomials over a field, there is a monic generator.

Reply:
I can redefine `ann_ideal_generator` to just `minpoly` if need be. Would it be just as good to show that the only choice of monic generator is the  `ann_ideal_generator`?
```lean
/-- If a monic generates the annihilating ideal, it must match our choice
 of the annihilating ideal generator. -/
lemma monic_generator_eq_minpoly (a : A) (p : 𝕜[X])
  (p_monic : p.monic) (p_gen : ideal.span {p} = ann_ideal 𝕜 a) :
  ann_ideal_generator 𝕜 a = p
```